File: goldenflow/transforms/categorical.py
from __future__ import annotations

import polars as pl

from goldenflow.transforms import register_transform

_TRUE_VALUES = {"yes", "y", "1", "true", "t"}
_FALSE_VALUES = {"no", "n", "0", "false", "f"}
_NULL_VALUES = {"n/a", "null", "none", "na", "nil", "nan", "-", ""}


@register_transform(
    name="boolean_normalize", input_types=["boolean", "string"], auto_apply=False, priority=50, mode="series"
)
def boolean_normalize(series: pl.Series) -> pl.Series:
    def _norm(val: str | None) -> bool | None:
        if val is None:
            return None
        v = val.strip().lower()
        if v in _TRUE_VALUES:
            return True
        if v in _FALSE_VALUES:
            return False
        return None

    return series.map_elements(_norm, return_dtype=pl.Boolean)


@register_transform(
    name="gender_standardize", input_types=["string"], auto_apply=False, priority=50, mode="series"
)
def gender_standardize(series: pl.Series) -> pl.Series:
    _map = {"male": "M", "m": "M", "female": "F", "f": "F"}

    def _std(val: str | None) -> str | None:
        if val is None:
            return None
        return _map.get(val.strip().lower(), val)

    return series.map_elements(_std, return_dtype=pl.Utf8)


@register_transform(
    name="null_standardize", input_types=["string"], auto_apply=True, priority=80, mode="series"
)
def null_standardize(series: pl.Series) -> pl.Series:
    def _std(val: str | None) -> str | None:
        if val is None:
            return None
        if val.strip().lower() in _NULL_VALUES:
            return None
        return val

    return series.map_elements(_std, return_dtype=pl.Utf8)


@register_transform(
    name="category_standardize",
    input_types=["string"],
    auto_apply=False,
    priority=45,
    mode="series",
)
def category_standardize(
    series: pl.Series, mapping: dict[str, list[str]] | None = None
) -> pl.Series:
    """Map variant values to canonical values. mapping: {canonical: [variant1, variant2, ...]}"""
    if not mapping:
        return series
    lookup: dict[str, str] = {}
    for canonical, variants in mapping.items():
        for v in variants:
            lookup[v.lower()] = canonical

    def _std(val: str | None) -> str | None:
        if val is None:
            return None
        return lookup.get(val.strip().lower(), val)

    return series.map_elements(_std, return_dtype=pl.Utf8)


@register_transform(
    name="category_from_file",
    input_types=["string"],
    auto_apply=False,
    priority=45,
    mode="series",
)
def category_from_file(
    series: pl.Series, lookup_path: str | None = None
) -> pl.Series:
    """Load mapping from a CSV/YAML file and standardize values.
    CSV must have columns: variant, canonical."""
    if not lookup_path:
        return series
    from pathlib import Path
    p = Path(lookup_path)
    if p.suffix == ".csv":
        import polars as pl_inner
        lookup_df = pl_inner.read_csv(p)
        mapping: dict[str, str] = {}
        for row in lookup_df.iter_rows(named=True):
            mapping[row["variant"].lower()] = row["canonical"]
    elif p.suffix in (".yaml", ".yml"):
        import yaml
        with open(p) as f:
            raw = yaml.safe_load(f) or {}
        mapping = {}
        for canonical, variants in raw.items():
            for v in variants:
                mapping[v.lower()] = canonical
    else:
        return series

    def _std(val: str | None) -> str | None:
        if val is None:
            return None
        return mapping.get(val.strip().lower(), val)

    return series.map_elements(_std, return_dtype=pl.Utf8)

File: packages/goldenflow-js/src/core/transforms/categorical.ts
/**
 * Categorical transforms — ported from goldenflow/transforms/categorical.py
 * Side-effect module: registers 5 categorical transforms on import.
 */

import type { ColumnValue } from "../types.js";
import { registerTransform } from "./registry.js";

// ---------------------------------------------------------------------------
// boolean_normalize (series, boolean|string, 50)
// ---------------------------------------------------------------------------

const TRUTHY = new Set(["yes", "y", "1", "true", "t"]);
const FALSY = new Set(["no", "n", "0", "false", "f"]);

function booleanNormalize(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const s = String(v).trim().toLowerCase();
    if (TRUTHY.has(s)) return true;
    if (FALSY.has(s)) return false;
    return v;
  });
}

registerTransform(
  { name: "boolean_normalize", inputTypes: ["boolean", "string"], priority: 50, mode: "series" },
  booleanNormalize,
);

// ---------------------------------------------------------------------------
// gender_standardize (series, string, 50)
// ---------------------------------------------------------------------------

function genderStandardize(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    if (typeof v !== "string") return v;
    const s = v.trim().toLowerCase();
    if (s === "male" || s === "m") return "M";
    if (s === "female" || s === "f") return "F";
    return v;
  });
}

registerTransform(
  { name: "gender_standardize", inputTypes: ["string"], priority: 50, mode: "series" },
  genderStandardize,
);

// ---------------------------------------------------------------------------
// null_standardize (series, string, 80, auto_apply)
// ---------------------------------------------------------------------------

const NULL_VARIANTS = new Set([
  "n/a", "null", "none", "na", "nil", "nan", "-", "",
]);

function nullStandardize(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    if (typeof v !== "string") return v;
    const s = v.trim().toLowerCase();
    if (NULL_VARIANTS.has(s)) return null;
    return v;
  });
}

registerTransform(
  { name: "null_standardize", inputTypes: ["string"], autoApply: true, priority: 80, mode: "series" },
  nullStandardize,
);

// ---------------------------------------------------------------------------
// category_standardize (series, string, 45, param: mapping=null)
// Mapping format: { canonical: [variant1, variant2, ...], ... }
// ---------------------------------------------------------------------------

function categoryStandardize(
  values: readonly ColumnValue[],
  mapping: unknown = null,
): ColumnValue[] {
  if (!mapping || typeof mapping !== "object") return values.slice();

  // Build a reverse lookup: lowercase variant -> canonical
  const lookup = new Map<string, string>();
  for (const [canonical, variants] of Object.entries(
    mapping as Record<string, string[]>,
  )) {
    if (Array.isArray(variants)) {
      for (const variant of variants) {
        lookup.set(String(variant).toLowerCase(), canonical);
      }
    }
    // Also map the canonical itself
    lookup.set(canonical.toLowerCase(), canonical);
  }

  return values.map((v) => {
    if (v === null) return null;
    if (typeof v !== "string") return v;
    const key = v.trim().toLowerCase();
    return lookup.get(key) ?? v;
  });
}

registerTransform(
  { name: "category_standardize", inputTypes: ["string"], priority: 45, mode: "series" },
  categoryStandardize,
);

// ---------------------------------------------------------------------------
// category_from_file (series, string, 45, param: lookup_path=null)
// Stub: Node-only implementation later; returns input unchanged.
// ---------------------------------------------------------------------------

function categoryFromFile(
  values: readonly ColumnValue[],
  lookupPath: unknown = null,
): ColumnValue[] {
  if (lookupPath) {
    console.warn("[goldenflow] category_from_file is not yet implemented in the JS port — returning values unchanged");
  }
  return values.slice();
}

registerTransform(
  { name: "category_from_file", inputTypes: ["string"], priority: 45, mode: "series" },
  categoryFromFile,
);
